trace-dispatcher/src/Cardano/Logging/ConfigurationParser.hs

```diff
@@ -95,57 +151,74 @@ parseRepresentation bs = transform (decodeEither' bs)
-
-instance AE.FromJSON ConfigRepresentation where
-    parseJSON (Object obj) = ConfigRepresentation
-                           <$> obj .: "TraceOptions"
-                           <*> obj .:? "TraceOptionForwarder"
-                           <*> obj .:? "TraceOptionNodeName"
-                           <*> obj .:? "TraceOptionPeerFrequency"
-                           <*> obj .:? "TraceOptionResourceFrequency"
-    parseJSON _ = mempty
+      -- | Convert from external to internal representation
+    toConfigOptions :: ConfigOptionRep -> [ConfigOption]
+    toConfigOptions ConfigOptionRep {..} =
+      case severity of
+        Nothing -> []
+        Just sev -> [ConfSeverity sev]
+      ++
+      case detail of
+        Nothing -> []
+        Just dtl -> [ConfDetail dtl]
+      ++
+      case backends of
+        Nothing -> []
+        Just bcks -> [ConfBackend bcks]
+      ++
+      case maxFrequency of
+        Nothing -> []
+        Just lim -> [ConfLimiter lim]
+
+
+-- | Convert from internal to external representation
+configToRepresentation :: TraceConfig -> ConfigRepresentation
+configToRepresentation traceConfig =
+     ConfigRepresentation
+        (toOptionRepresentation (tcOptions traceConfig))
+        (tcForwarder traceConfig)
+        (tcNodeName traceConfig)
+        (tcPeerFrequency traceConfig)
+        (tcResourceFrequency traceConfig)
+  where
+    toOptionRepresentation :: Map.Map [Text] [ConfigOption]
+                              ->  Map.Map Text ConfigOptionRep
+    toOptionRepresentation internalOptMap =
+      foldl' conversion Map.empty (Map.toList internalOptMap)
+
+    conversion :: Map.Map Text ConfigOptionRep
+                -> ([Text],[ConfigOption])
+                -> Map.Map Text ConfigOptionRep
+    conversion accuMap (ns, options) =
+      let nssingle = intercalate "." ns
+          optionRep = fromOptions options
+      in  Map.insert nssingle optionRep accuMap
+
+    fromOptions :: [ConfigOption] -> ConfigOptionRep
+    fromOptions opts =
+      ConfigOptionRep
+      { severity = case filter (\case
+                          ConfSeverity _ -> True
+                          _ -> False) opts of
+                      ConfSeverity sev : _ -> Just sev
+                      _ -> Nothing
+      , detail   = case filter (\case
+                          ConfDetail _ -> True
+                          _ -> False) opts of
+                      ConfDetail det : _ -> Just det
+                      _ -> Nothing
+      , backends = case filter (\case
+                          ConfBackend _ -> True
+                          _ -> False) opts of
+                      ConfBackend back : _ -> Just back
+                      _ -> Nothing
+      , maxFrequency = case filter (\case
+                          ConfLimiter _ -> True
+                          _ -> False) opts of
+                      ConfLimiter freq : _ -> Just freq
+                      _ -> Nothing
+      }
```

Comment:
Could one use `listToMaybe [d | ConfDetail d <- opts]` here?

Reply:
toOptionRepresentation looks like Map.mapKeys (intercalate ".") . Map.map fromOptions could avoid needing the helper function, if you like the idea. (Map.mapKeysMonotonic might be applicable too.)

Reply:
@mgmeier changed. @NadiaYvette sorry, but I like my simple fold, which is easier for me to understand